app/assets/javascripts/tictactoe.js

```diff
@@ -1 +1,203 @@
 // Code your JavaScript / jQuery solution here
+var turn;
+var gameId;
+$(document).ready(initialize)
+
+function initialize() {
+  attachListeners();
+  turn = 0;
+  gameId = undefined;
+}
+
+function player() { // X=even, Y=odd
+  if ((turn % 2) === 0) {
+    return "X"; 
+  } else {
+    return "O";
+  }
+}
+
+function updateState(el) { // Pass an element to update it with the current player's symbol
+  const x = el.attributes[0].nodeValue; // Get x coordinate (0-2)
+  const y = el.attributes[1].nodeValue; // Get y coordinate (0-2)
+  let square = $(`[data-x=${x}][data-y=${y}]`);
+
+  if (square.text() === "") {
+    square.text(player());
+    return true;
+  } else {
+    return false;
+  }
+}
+
+function setMessage(string) {
+  $("#message").text(string);
+}
+
+function checkWinner() {
+  let xSpots = [];
+  let o_spots = [];
+  let win = [
+    [0,1,2],[0,3,6],[0,4,8],[6,4,2],[2,5,8],[6,7,8],[3,4,5],[1,4,7]
+  ];
+  let square;
+
+  for (let x = 0; x < 3; x++) {
+    for (let y = 0; y < 3; y++) {
+      square = $(`[data-x=${x}][data-y=${y}]`).text()
+
+      if (square === "X") {
+        xSpots.push(x + y*3);
+      } else if (square === "O") {
+        o_spots.push(x + y*3);
+      }
+    }
+  }
+
+  // Check each player's spots for a win combination
+  for (let i = 0; i < win.length; i++) {
+    if (xSpots.includes(win[i][0]) 
+      && xSpots.includes(win[i][1]) 
+      && xSpots.includes(win[i][2])) {
+        setMessage('Player X Won!');
+        return true;  
+    } else if (o_spots.includes(win[i][0]) 
+      && o_spots.includes(win[i][1]) 
+      && o_spots.includes(win[i][2])) {
+        setMessage('Player O Won!');
+        return true;
+    } 
+  }
+  // -- code will only get here if no player has won ---
+  // If board is full, it's a tie
+  if (fullBoard()) {
+    setMessage('Tie game.');
+  }
+  return false;
+}
+
+function doTurn(el) {
+  if (updateState(el)) {
+    turn += 1;
+  }
+  if (checkWinner()) {
+    saveGame();
+    resetGame();
+  } else if (fullBoard()) {
+    // Must be a draw
+    saveGame();
+    resetGame();
+  }
+  /*
+  updateState(el);
+  turn++;
+  */
+}
+
+function attachListeners() {
```

Comment:
It looks like you are calling `attachListeners` in your `document.ready`, but you have 1 or more listeners outside of that function.

```javascript
$(document).ready(function() {
  attachListeners();
})

function attachListeners() {
  $('td').on('click', function() {
    if (!$.text(this) && !checkWinner()) {
      doTurn(this);
    }
  });

  $('#save').on('click', () => saveGame());
  $('#previous').on('click', () => showPreviousGames());
  $('#clear').on('click', () => resetBoard());
}
```